Log centralization design (#2668)
scality/metalk8s: docs/developer/architecture/logs.rst

log centralization service is available when deploying/upgrading MetalK8s.

Components
----------

Our Log Centralization system can be splitted into several components as
follows:

.. parsed-literal::

  +-----------+     +-------------+     +----------+     +---------+
  | Collector | ==> | Distributor | ==> | Ingester | <=> | Querier |
  +-----------+     +-------------+     +----------+     +---------+

.. todo:: do a better schema. :)

Collector
~~~~~~~~~

The collector is responsible for processing logs from all the sources
(files, journal, containers, ...), enriching the logs with metadata (labels)
coming from parsing/filtering them or by querying external sources such as
APIs, then finally forwarding these logs to one or multiple distributors.

Distributor (Router)
~~~~~~~~~~~~~~~~~~~~

The distributor is the component that receives incoming streams from the
collectors, validates and prepares the streams for storage, then forwards
them to the right ingester.
The distributor can also do some processing on the streams to enrich them
with metadata (labels).

Ingester (Storage)
~~~~~~~~~~~~~~~~~~

The job of the ingester is to write log streams on persistent storage such

> as a filesystem or a database, but also to return the data requested by

Reviewer: in Loki ecosystem Ingesters only return data that have not been flashed to storage they re not used to read the data back,no?

Author: true, it's not clear here, even if this part of the documentation is not directly related to Loki, but more a general concept.

Design Choices
--------------

To choose which solution fits the most our needs, we did a benchmark of some
of the solutions to compare their performances.
Our choice for the final design has been greatly motivated by these numbers,

Check `this document`_ for details.

.. _this document: https://docs.google.com/spreadsheets/d/1f2Pz9zBbkDzpHr8KzST3nTcfvWJqgoNcCEd0-4WIeow/edit#gid=0

.. todo::

  export the spreadsheet and include it or do directly a quick recap tab here.

Fluent Bit + Fluentd (optional) + Loki
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Fluent Bit
""""""""""

We choose Fluent Bit as the collector because it allows to scrape all the logs
we need (journal & containers), enriches them with the Kubernetes API and it
has a very low resources footprint.

Moreover, it supports multiple distributors such as Loki, ES, Splunk, etc. at
the same time, which is a very important point if a user also wants to forward
the logs to an external log centralization system (e.g. long term archiving).

Check here_ for the official list of supported outputs.

.. _here: https://docs.fluentbit.io/manual/pipeline/outputs

Loki
""""

> Loki has been choosen as the distributor, ingester & querier because, like

Reviewer: chosen

Fluent Bit, it has a really small impact on resources and is very cost
effective regarding storage needs.

It also uses the LogQL syntax for queries, which is pretty close to what
we already have with Prometheus and PromQL, so it eases the integration
in our tools.

Fluentd (optional)
""""""""""""""""""

> Fluentd could be optionally added between Fluent Bit and Loki to offers

Reviewer: offer

Loki
""""

Loki will be deployed as a StatefulSet as it needs persistent volumes to write
the logs. It will be running only on control plane nodes and we need at least
2 replicas (except on a single node platform) of it to ensure its high
availability.

Loki HA
_______

Since we're using filesystem to store Loki's data, we have basically 2 ways
for having multiple Loki instances running at the same time, with the same
set of data.

Either we do like Prometheus and we store everything on every instance of
Loki, but it means we raise the storage, RAM and CPU needs for each additionnal
instance, either we use a `new experimental feature`_ from Loki 1.5.0 where
Loki ingesters use a hash ring and talk between them to route the queries to
the right one. This approach needs an external KV store to work such as
etcd or Consul.

.. _new experimental feature: https://github.com/grafana/loki/blob/master/docs/operations/storage/filesystem.md#new-and-very-experimental-in-150-horizontal-scaling-of-the-filesystem-store

.. note::

  What do we do ? Do we use metalk8s etcd to store loki ring ? :/

  Or do we add a new instance of etcd or consul ?

>   This feature is marked as experimental, should we even use it ?

Reviewer: Looks very prone to changes if you ask me :/

Author: I agree, and probably bugged too.
But otherwise we can only do "replication" like Prometheus, meaning we have to send all streams on all Loki instances with the overhead of resources it implies.
Maybe we can go for the dirty solution and then if it becomes stable at some point, implement this.
But, to me, the goal of using Loki is to reduce storage (and also RAM & CPU) needs, if we go this way, it'll obviously not be the best.

Reviewer 2: Since the Loki part of MetalK8s is only here for short term retention to me it's ok to store everything on every instance of Loki.

Reviewer 3: As discussed offline, we want to consider our infra services as simple _poor man's solutions_, and document how to get metrics/logs/alerts/... out of the cluster into other systems (e.g. Splunk for logs).

Author: Ok, good to me, just wanted to be sure everyone is aware and agree on this.

A Grafana dashboard gathering health/performance KPIs, as well as alerts for
log centralization service is available when deploying/upgrading MetalK8s.

Components
----------

> Our Log Centralization system can be splitted into several components as

Reviewer 3: "split", not "splitted"

follows:

.. parsed-literal::

  +-----------+     +-------------+     +----------+     +---------+
  | Collector | ==> | Distributor | ==> | Ingester | <=> | Querier |
  +-----------+     +-------------+     +----------+     +---------+

> .. todo:: do a better schema. :)

Reviewer 3: What would you add? The only thing missing from the schema, IMHO, is a way to see how these components can scale (e.g. if I have multiple collectors, do they all send to a single distributor? or do they each have their own? something in between?)

Author: Yes, that's my point, because here we don't really understand interactions when there is multiple Loki instances, for example, or do each collector write to one or multiple distributors ?

Collector
~~~~~~~~~

The collector is responsible for processing logs from all the sources
(files, journal, containers, ...), enriching the logs with metadata (labels)

> coming from parsing/filtering them or by querying external sources such as

Reviewer 3: Or coming from context (the source itself can provide metadata: e.g. the file path, container name and parent pod...)

APIs, then finally forwarding these logs to one or multiple distributors.

Distributor (Router)
~~~~~~~~~~~~~~~~~~~~

The distributor is the component that receives incoming streams from the

> collectors, validates and prepares the streams for storage, then forwards

Reviewer 3: What do we mean by validation and preparation of log streams? Is the distributor responsible for chunking/formatting? I thought that was the ingester's job.

Author: That's not well written, it does not chunk or format anything, just checking that everything in the stream is valid, it means:
- The metric labels name are formally correct
- The configured max number of labels per metric is respected
- The configured max length of a label name and value is respected
- The timestamp is not older/newer than the configured min/max time range

them to the right ingester.

> The distributor can also do some processing on the streams to enrich them

Reviewer 3: Can't a distributor have other functions, like filtering or buffering?

Author: Yes, that's "some processing".
Maybe it should be clearer.

with metadata (labels).

Ingester (Storage)
~~~~~~~~~~~~~~~~~~

The job of the ingester is to write log streams on persistent storage such
as a filesystem or a database, but also to return the data requested by

> the queriers, if still in memory.

Reviewer 3: To make it more obvious, the ingester serves as a buffer between distributor and storage, because writing large (potentially compressed) chunks of data (potentially over the network) is more efficient than writing each event individually as it arrives. As such, a querier may need to ask an ingester about what is in the buffer.

Querier
~~~~~~~

The querier interprets the queries it receives from clients and then asks the

> ingesters for the corresponding data, then fallback on storage backend if not

Reviewer 3: Is it a mandatory behavior? Can't a querier ignore ingesters entirely, focusing on querying the storage? How does a querier know about ingesters? What if the ingester is busy, will the query wait? Isn't this pattern encouraging tight coupling between querier and ingester?

Author: It's the default behavior of Loki, depending on the configuration and the logging rate, a log could stay in ingester buffer for few minutes before being flushed to the storage backend, so if you do not query the ingester, you will not get the latest logs.
If the ingester is busy, the querier will wait `querier_config.query_timeout` time (1m by default) before timing out.
We can disable this mechanism or at least minimize it, with the option `quierer_config.query_ingesters_within`, which represents the maximum lookback beyond which queries are not sent to ingester (default to 0s, which means it always query ingesters, before falling back to storage).
Maybe we can set it to `-1` to disable it.

present in memory and returns it to the clients.

> It also takes care of deduplication of data because of the replication.

Reviewer 3: Can it aggregate results from multiple storage backends? Imagine: I'm configuring my ingesters to save all logs on filesystem, with 1 week retention, and to also save error logs in an external Cassandra DB, with 2 months retention. Is this feasible? How do queries behave in this case?

Author: It does not seem to be possible, AFAICT you can only have 1 storage backend.

Design Choices
--------------

> To choose which solution fits the most our needs, we did a benchmark of some

Reviewer 3: "fits best" i.s.o. "fits the most"

of the solutions to compare their performances.

> Our choice for the final design has been greatly motivated by these numbers,

Reviewer 3: Ending with a comma, the suspense is too damn high! ;)

> Check `this document`_ for details.

Reviewer 3: FWIW, you can inline links using the following syntax:

```rst
`my link text <link_url>`_
```

Author: Yes, I did it on purpose, for consistency, we do it this way in most of the documentation.

Reviewer 3: But then the `this document` reference is global to the whole page... Doesn't make sense for this kind of link IMHO.

> .. _this document: https://docs.google.com/spreadsheets/d/1f2Pz9zBbkDzpHr8KzST3nTcfvWJqgoNcCEd0-4WIeow/edit#gid=0

Reviewer 3: Private docs in a public repo? Could be worth copying over the results, either as some CSV or directly in a RST table.

(Edit: just read the to-do below, let's have it done before merging please :smirk:)

Author: Yes, it's just there for the reviewers to understand what will be put here.

.. todo::

  export the spreadsheet and include it or do directly a quick recap tab here.

> Fluent Bit + Fluentd (optional) + Loki

Reviewer 3: Let's have a small schema showing the role of each entity (collector/distributor/...) and how we imagine them being scheduled in a cluster.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Fluent Bit
""""""""""

We choose Fluent Bit as the collector because it allows to scrape all the logs

> we need (journal & containers), enriches them with the Kubernetes API and it

Reviewer 3: Is it using the API? I thought it was only reading local files.

Author: Yes, it's reading the Kubernetes API to enrich the logs with metadata such as container labels.

has a very low resources footprint.

> Moreover, it supports multiple distributors such as Loki, ES, Splunk, etc. at

Reviewer 3: Distributors?

Author: Backend would probably be more appropriate.

the same time, which is a very important point if a user also wants to forward
the logs to an external log centralization system (e.g. long term archiving).

Check here_ for the official list of supported outputs.

.. _here: https://docs.fluentbit.io/manual/pipeline/outputs

Loki
""""

Loki has been chosen as the distributor, ingester & querier because, like
Fluent Bit, it has a really small impact on resources and is very cost
effective regarding storage needs.

It also uses the LogQL syntax for queries, which is pretty close to what
we already have with Prometheus and PromQL, so it eases the integration
in our tools.

Fluentd (optional)
""""""""""""""""""

Fluentd could be optionally added between Fluent Bit and Loki to offer

> more plugins and reduce the impact of Fluent Bit on workload nodes.

Reviewer 3: Let's try to be slightly more explicit:
- more plugins are great, but what for? (For customers trying to push logs in some exotic backend? Some other parsing feature Fluent Bit is missing?)
- reduce the impact how? (IIUC, by removing the need from Fluent Bit to parse log events. Is it still collecting metadata from context?)

It's not recommended for small platforms (few nodes) as it adds a big overhead
in resources consumption.

Rejected Design Choices
-----------------------

Promtail + Loki
~~~~~~~~~~~~~~~

Promtail has been rejected because, even if it consumes very few resources,
it can only integrate with Loki and we need something more versatile, with
the ability to interact with different distributors.

Fluentd + Loki
~~~~~~~~~~~~~~

This architecture has been rejected because it means we need 1 fluentd instance
per node, which increases a lot the resources consumption compared to other
solutions.

Logstash + Elasticsearch
~~~~~~~~~~~~~~~~~~~~~~~~

This architecture is probably the most common one, but it has not been taken
into consideration because we want to focus on having the minimum resources
consumption and these components can really hog RAM & CPU. Beats could be used
as the log collector to reduce the impact of Logstash, but Elasticsearch still

> consumes a lot of resources.

Reviewer 3: Well, it does a bunch more than Loki as well: distributed storage, full-text indexing...
Given our requirements, we especially want the ingestion load to be small, and that's where Loki shines, while we don't care as much about querying (and that's where ES is strong).

Author: Yes, I didn't write it because, we say in the document that our focus is on resources consumption.
So, if we want something light, it often comes with less functionalities.
I could still precise it.

Implementation Details
----------------------

Deployment
~~~~~~~~~~

All the components will be deployed using Kubernetes manifests through Salt
inside a ``metalk8s-logging`` Namespace.

> As we already did for other components (e.g. Prometheus Operator), the

Reviewer 3: Why not use the Logging Operator from Banzai Cloud?

Author: Well, I've considered it, but you can't do a simple architecture with fluentbit + loki (without fluentd).
Otherwise it looks like it's pretty good stuff and would ease many things.
But, if we consider the deployment of fluent as not optional, we could use it.

Reviewer 3: Would need empirical confirmation, but as @xaf-scality suggested, we may be able to configure fluentd as passthrough, reducing its impact on resources.

manifests for Loki and Fluent Bit will be generated statically by rendering
the `loki-stack helm chart`_::

  helm repo add loki https://grafana.github.io/loki/charts
  helm repo update
  helm fetch -d charts --untar loki/loki-stack
  ./charts/render.py loki-stack --namespace metalk8s-logging \
    charts/loki-stack.yaml charts/loki-stack/ \
    > salt/metalk8s/addons/loki-stack/deployed/charts.sls

.. _loki-stack helm chart: https://github.com/grafana/loki/tree/master/production/helm

We will use the same approach for Fluentd, using the `fluentd helm chart`_::

  helm fetch -d charts --untar stable/fluentd
  ./charts/render.py fluentd --namespace metalk8s-logging charts/fluentd.yaml \
    charts/fluentd/ > salt/metalk8s/addons/fluentd/deployed/charts.sls

.. _fluentd helm chart: https://hub.helm.sh/charts/stable/fluentd

> Container images must be added in the buildchain

Reviewer 3: Not sure we need this level of detail in the design doc, but will be highly appreciated in implementation tickets ;)

Author: removed

``buildchain/buildchain/image.py`` and then replaced in
``charts/loki-stack.yaml`` and ``charts/fluentd.yaml`` using the
``__image__()`` macro to be replaced by jinja syntax when rendering the
charts and then dynamically computed at run time.

Fluent Bit
""""""""""

Fluent Bit will be deployed as a DaemonSet, because we need one collector on

> each control or workload plane node to be able to collect logs from both the

Reviewer 3: No need for distinction between control and workload plane.

> Kubernetes platform and the applications that run on top of it.

Reviewer 3: And all the system stuff running alongside K8s ;) (e.g. Salt minion)

Loki
""""

> Loki will be deployed as a StatefulSet as it needs persistent volumes to write

Reviewer 4: Don't we need specific labels when provisioning PVs for Loki? 
Do we recommend having dedicated disks or can we use LVs? do we need SSD or HDD is okay? Can we easily increase PV capacity to handle more logs?

Reviewer 3: I don't think we support resizing PVs. Also, for the question of labels, it's really an implementation detail. The design would need to clarify the sizing rule, however (depending on retention + ingestion rate).

> Loki will be deployed as a StatefulSet as it needs persistent volumes to write

Reviewer 3: So we can't split Loki components into different K8s objects and scale them independently? What about autoscaling?

Author: Not possible to split components using filesystem as storage backend, which means no autoscaling either because the monolithic pod is stateful thank to the ingester volume.

Reviewer 3: Can't split and horizontally scale - true
Can't autoscale - debatable: we may want to adjust `memory` and `cpu` consumption from the querier container (through vertical **autoscaling** :wink:)

the logs. It will be running only on infra nodes and we need at least
2 replicas (except on a single node platform) of it to ensure its high
availability.

Loki HA
_______

Since we're using filesystem to store Loki's data, we have basically 2 ways
for having multiple Loki instances running at the same time, with the same
set of data.

Either we do like Prometheus and we store everything on every instance of
Loki, but it means we raise the storage, RAM and CPU needs for each additionnal
instance, either we use a `new experimental feature`_ from Loki 1.5.0 where

> Loki ingesters use a hash ring and talk between them to route the queries to

Reviewer 3: Then it's not about HA, but distributed storage, right? If I lose one, I lose data? Or is it doing some form of erasure coding?

the right one. This approach needs an external KV store to work such as
etcd or Consul.

We choose to use the first approach as the second is not production ready
and since we plan to only do short term retention on Loki, the impact on
storage will not be that much important.
Moreover, it eases the deployment and maintenance since there is not an
extra component.

.. _new experimental feature: https://github.com/grafana/loki/blob/master/docs/operations/storage/filesystem.md#new-and-very-experimental-in-150-horizontal-scaling-of-the-filesystem-store

Fluentd
"""""""

Fluentd will be deployed as a Deployment with constraints to schedule it only
on infra nodes.

> By default we need at least 2 replicas (except on a single node platform) to be

Reviewer 3: How are Fluent Bits forwarding the logs in this case? To each available Fluentd? To the corresponding Service, hoping all logs will end up in Loki somehow?

sure we always have at least one instance of Fluentd running.

Configuration
~~~~~~~~~~~~~

Fluent Bit
""""""""""

Fluent Bit needs to be configured to scrape and handle properly journal and
containers logs by default.

> Its configuration

Reviewer 3: ?

Author: Looks like I didn't finish my sentence. :D

For containers logs, we want to add the following labels:

 - node: the node it comes from
 - namespace: the namespace the pod is running in
 - instance: the name of the pod
 - container: the name of the container

For journal we want these labels:

 - node: the host it comes from
 - unit: the name of the unit generating these logs

.. note::

  What if host != node ?

>   Do we want more labels (e.g. "app", "environment") ?

Reviewer 4: since we have namespace and container, I would say it is sufficient for a first implem

Reviewer 3: Furthermore, these aren't always set. I would still try to get a value for `app.kubernetes.io/name` if available, with a placeholder if label was not set, because we have no other equivalent.

>   How do we allow Solutions to add filters/parsers (not in the reqs) ?

Reviewer 4: Adding filters means, adding label to the container logs?
Not sure to understand what means adding parser?

Reviewer 3: Filters are rules to drop log events.
Parsers are basically regular expressions used to extract labels from log events.

Author: Filters are also used to redirect the inputs in the right parsers, but yes that's basically the idea.

Reviewer 4: So that is a very interesting feature and it should be part of the requirements (to be discussed during today's workshop)

>   How do we allow Solutions to add filters/parsers (not in the reqs) ?

Reviewer 3: Definitely required, not only for Solutions (e.g. we add a component, we need a new parser for its log format, what do we do?).

Again, why not consider the Logging Operator? It defines all of these in CRs, making it easy to customize and extend.

add new routes to push the log streams to, which will be handled by the CSC
mechanisms, using the ``metalk8s-fluent-bit-config`` ConfigMap in the
``metalk8s-logging`` Namespace.

The ConfigMap will look like the followings:

.. code-block:: yaml

  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: metalk8s-fluent-bit-config
    namespace: metalk8s-logging
  data:
    config.yaml: |-
      apiVersion: addons.metalk8s.scality.com
      kind: FluentBitConfig
      spec:
        output: |-
          [OUTPUT]
            Name            es
            Match           *
            Host            localhost
            Port            9200
            Logstash_Format On
            Replace_Dots    On
            Retry_Limit     False

.. note::

  Fluent Bit configuration is not a YAML file and looks more like an INI one.
  CSC is not able to take care of these files and even if we add support for
  this format, it will be hard to implement the merging with defaults mechanism
  as blocks can not really be identified, because there is no unique name or ID.
  Since we only need to provide to the user the possiblity to add a new output,

>   it should be sufficient as we don't need to edit what's deployed by MetalK8s.

Reviewer 3: What about switching between with and without Fluentd?

Loki
""""

Loki's configuration is stored as a Secret, as for Fluent Bit, we need
to expose few parameters to the user for customization.
We will use the CSC mechanism to provide the interface for customization,
with a ConfigMap ``metalk8s-loki-config`` in ``metalk8s-logging`` Namespace.

A Salt module will be developed to manipulate this object, so the logic
can be kept in only one place.

This module must provide necessary methods to show or edit the configuration
in 2 different ways:

   * simple
   * advanced

The ``simple`` mode will only display and allow to change some specific

> configuration, such as the retention or the number of replicas, and in an as

Reviewer 3: I'm still worried about scaling Loki components independently from one another... Would this fit a "simple" mode?

> simple as possible manner for the user.

Reviewer 3: "the simplest possible form"

The ``advanced`` mode will allow to change all the configuration points,
exposing the whole configuration as a plain YAML.

This module will then be exposed through a CLI and a UI.

Fluentd
"""""""

Monitoring
~~~~~~~~~~

Prometheus
""""""""""

To monitor every services in our log centralization stack, we will need to
deploy ``ServiceMonitor`` object and expose the ``/metrics`` route of all
these components.
It will allow Prometheus Operator to configure Prometheus and automatically
start scraping these services.
This can be achieved by adding the following configuration in the helm
chart values:

For Fluent Bit, in ``charts/loki-stack.yaml``:

.. code-block:: yaml

  fluent-bit:
    serviceMonitor:
      enabled: true
      additionalLabels:
        release: "prometheus-operator"

For Loki, in ``charts/loki-stack.yaml``:

.. code-block:: yaml

  loki:
    serviceMonitor:
      enabled: true
      additionalLabels:
        release: "prometheus-operator"

For Fluentd, in ``charts/fluentd.yaml``:

.. code-block:: yaml

  metrics:
    serviceMonitor:
      enabled: true
      additionalLabels:
        release: "prometheus-operator"

> We also need to define alert rules based on metrics exposed by these services.

Reviewer 4: what would be the formula to compute Log centralization Health?

Author: It's written, we need to define this formula.
We could have a basic formula which only make sure that every pods (or at least the daemonset ones & 1 per deployment) in the metalk8s-logging namespace is healthy, what do you think ?

Reviewer 3: Suggestion from @xaf-scality was to also consider the whole pipeline: can we use alerting rules to compare the amount of logs seen by collectors with what gets ingested in Loki? Would help us see if something in the pipeline is broken, meaning we'd be losing logs.

        type: loki
        access: proxy
        url: http://loki.loki.svc.cluster.local:3100
        version: 1

To read the log we will also need a dashboard, adding a ConfigMap
``loki-logs-dashboard`` in Namespace ``metalk8s-monitoring``:

.. code-block:: yaml

  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: loki-logs-dashboard
    namespace: metalk8s-monitoring
    labels:
      grafana_dashboard: "1"
  data:
    loki-logs.json: [TO BE DEFINED]

.. note::

  The ``grafana_datasource: "1"`` and ``grafana_dashboard: "1"`` labels are
  what is used by the Prometheus Operator to retrieve datasource and dashboard
  for Grafana.

Iteration 1
~~~~~~~~~~~

The goal is to have a working log centralization system, with logs accessible
from Grafana:

* Deployment of the stack (without Fluentd)
* Deploy Grafana datasource & dashboard
* Document the log centralization system
* Simple pre-merge test to ensure the log pipeline is working
* Customization of Loki, Fluent Bit and Fluentd (editing the CSC ConfigMap)

Iteration 2
~~~~~~~~~~~

* Deployment of Fluentd
* Customization of Loki, Fluent Bit and Fluentd (simple + advanced modes)
* Document how to configure these components

> * Define monitoring rules

Reviewer 4: and display in MetalK8s Overview page

Reviewer 3: Not sure if really needed in an Overview page, as it's not a critical component of the system. I'd rather have dedicated pages for the various "infra services" we deploy (auth, monitoring, logging).

Fluent Bit
""""""""""

We choose Fluent Bit as the collector because it allows to scrape all the logs
we need (journal & containers), enriches them with the Kubernetes API and it
has a very low resources footprint.

Moreover, it supports multiple distributors such as Loki, ES, Splunk, etc. at
the same time, which is a very important point if a user also wants to forward
the logs to an external log centralization system (e.g. long term archiving).

Check here_ for the official list of supported outputs.

.. _here: https://docs.fluentbit.io/manual/pipeline/outputs

Loki
""""

Loki has been chosen as the distributor, ingester & querier because, like
Fluent Bit, it has a really small impact on resources and is very cost
effective regarding storage needs.

It also uses the LogQL syntax for queries, which is pretty close to what
we already have with Prometheus and PromQL, so it eases the integration
in our tools.

Fluentd (optional)
""""""""""""""""""

Fluentd could be optionally added between Fluent Bit and Loki to offer
more plugins and reduce the impact of Fluent Bit on workload nodes.

It's not recommended for small platforms (few nodes) as it adds a big overhead

> in resources consumption.

Reviewer 3: It's not clear what role Fluentd would play out in this case: Distributor, or a part of the Collector pipeline with Fluentbit?

For the "optional" part of this, is the design stating our implementation may include Fluentd, or is it saying we will have both options available in the product and expose this to our users?

Author: Oh yes, I agree, that's not crystal clear.
Fluentd will be part of the Collector pipeline, I'll explain it.

It means we will have both options available, depending on the needs, I will also try to explain a bit more.

Author: Fluentd has been removed from the design.

    charts/loki-stack.yaml charts/loki-stack/ \
    > salt/metalk8s/addons/loki-stack/deployed/charts.sls

.. _loki-stack helm chart: https://github.com/grafana/loki/tree/master/production/helm

We will use the same approach for Fluentd, using the `fluentd helm chart`_::

  helm fetch -d charts --untar stable/fluentd
  ./charts/render.py fluentd --namespace metalk8s-logging charts/fluentd.yaml \
    charts/fluentd/ > salt/metalk8s/addons/fluentd/deployed/charts.sls

.. _fluentd helm chart: https://hub.helm.sh/charts/stable/fluentd

Container images must be added in the buildchain
``buildchain/buildchain/image.py`` and then replaced in
``charts/loki-stack.yaml`` and ``charts/fluentd.yaml`` using the
``__image__()`` macro to be replaced by jinja syntax when rendering the
charts and then dynamically computed at run time.

Fluent Bit
""""""""""

Fluent Bit will be deployed as a DaemonSet, because we need one collector on
each control or workload plane node to be able to collect logs from both the
Kubernetes platform and the applications that run on top of it.

Loki
""""

Loki will be deployed as a StatefulSet as it needs persistent volumes to write

> the logs. It will be running only on control plane nodes and we need at least

Reviewer 2: To me loki will be stuck to infra nodes not control plane

```suggestion
the logs. It will be running only on infra nodes and we need at least
```

Author: true.

2 replicas (except on a single node platform) of it to ensure its high
availability.

Loki HA
_______

Since we're using filesystem to store Loki's data, we have basically 2 ways
for having multiple Loki instances running at the same time, with the same
set of data.

Either we do like Prometheus and we store everything on every instance of
Loki, but it means we raise the storage, RAM and CPU needs for each additionnal
instance, either we use a `new experimental feature`_ from Loki 1.5.0 where
Loki ingesters use a hash ring and talk between them to route the queries to
the right one. This approach needs an external KV store to work such as
etcd or Consul.

.. _new experimental feature: https://github.com/grafana/loki/blob/master/docs/operations/storage/filesystem.md#new-and-very-experimental-in-150-horizontal-scaling-of-the-filesystem-store

.. note::

  What do we do ? Do we use metalk8s etcd to store loki ring ? :/

  Or do we add a new instance of etcd or consul ?

  This feature is marked as experimental, should we even use it ?

Fluentd
"""""""

Fluentd will be deployed as a Deployment with constraints to schedule it only

> on control plane nodes.

Reviewer 2: Same as Loki to me it's on infra nodes

Author: Yes

By default we need at least 2 replicas (except on a single node platform) to be
sure we always have at least one instance of Fluentd running.

Configuration
~~~~~~~~~~~~~

Fluent Bit
""""""""""

Fluent Bit needs to be configured to scrape and handle properly journal and
containers logs by default.
Its configuration

For containers logs, we want to add the following labels:

 - node: the node it comes from
 - namespace: the namespace the pod is running in
 - instance: the name of the pod
 - container: the name of the container

For journal we want these labels:

 - node: the host it comes from
 - unit: the name of the unit generating these logs

.. note::

>   What if host != node ?

Reviewer 2: To me we may want to have node name everywhere and not hostname (or maybe having both :thinking:) , no ?

Author: Yes, I mean we want to be able to filter with a single name to get both logs from journal and k8s, but the hostname is not necessarily equal to the node name.
Maybe we can find a way to enrich the containers logs with the hostname.

Reviewer 3: A fluentbit is tied to a single Node, should be easy for it to add the Node name everywhere, no?

   | promtail             | 975              | 928                |
   +----------------------+------------------+--------------------+
   | fluent-bit           | 790              | 830                |
   +----------------------+------------------+--------------------+
   | fluent-bit + fluentd | 1311             | 1967               |
   +----------------------+------------------+--------------------+

With 10k events/sec, composed of 1000 distinct streams:

   +----------------------+------------------+--------------------+
   |                      | **CPU avg in m** | **RAM avg in MiB** |
   +----------------------+------------------+--------------------+
   | promtail             | 1292             | 1925               |
   +----------------------+------------------+--------------------+
   | fluent-bit           | 1040             | 834                |
   +----------------------+------------------+--------------------+
   | fluent-bit + fluentd | 1447             | 1902               |
   +----------------------+------------------+--------------------+

We can see that the Fluent Bit + Loki architecture has the smallest impact

> on resources.

Reviewer 2: To me, we can also see that fluent-bit + fluentd seems to scale much better than fluent-bit alone (and fluent-bit always have impact on workload where fluentd could be on dedicated infra nodes)
Just saying, I don't say we should go for fluent-bit + fluentd but... interesting (may be worth to mention this somewhere)

Author: True, I can add one or two sentences explaining this.

Reviewer 3: Also, we're only comparing scaling in number of log streams here (which would, in theory, mostly scale with the number of nodes). We also need to consider scaling in ingestion rate (which can happen for uncontrolled reasons like a spike in a workload API usage).

Finally, even though you showed the overall CPU/RAM consumption, there is value in seeing which component consumes what (especially if we want to evaluate the impact of fluent-bit on each node).

This document should clearly state how the proposed solutions compare (maybe ignoring Promtail for limited configurability, which is explained below :ok_hand:), for future readers / users of MetalK8s to understand our choices and future strategy.

has a very low resources footprint.

Moreover, it supports multiple backend such as Loki, ES, Splunk, etc. at
the same time, which is a very important point if a user also wants to forward
the logs to an external log centralization system (e.g. long term archiving).

Check `here <https://docs.fluentbit.io/manual/pipeline/outputs>`__
for the official list of supported outputs.

Loki
""""

Loki has been chosen as the distributor, ingester & querier because, like
Fluent Bit, it has a really small impact on resources and is very cost
effective regarding storage needs.

It also uses the LogQL syntax for queries, which is pretty close to what
we already have with Prometheus and PromQL, so it eases the integration
in our tools.

Rejected Design Choices
-----------------------

Logging Operator
~~~~~~~~~~~~~~~~

`Logging Operator`_ seemed to be a good candidate for the implementation
we choose, offering the ability to deploy and configure Fluent Bit and Fluentd,
but Fluentd is not optional and seems to have a central place as most of the
parsing/filtering is done by this one, which means a bigger footprint on the
hardware resources.

> .. _Logging Operator: https://github.com/banzaicloud/logging-operator

Reviewer 2: Minor: Maybe you should move this one just after `Fluent-bit + Fluentd + Loki` section

Author: Why not.

an extra Fluentd means more resources consumption and add unnecessary
complexity in the log centralization stack.

Logstash + Elasticsearch
~~~~~~~~~~~~~~~~~~~~~~~~

This architecture is probably the most common one, but it has not been taken
into consideration because we want to focus on having the minimum resources
consumption and these components can really hog RAM & CPU. Beats could be used
as the log collector to reduce the impact of Logstash, but Elasticsearch still
consumes a lot of resources.
Even if this solution offers a lot of powerful functionalities
(e.g. distributed storage, full-text indexing), we don't really need them and
want to focus on the smallest hardware footprint.

Implementation Details
----------------------

Deployment
~~~~~~~~~~

All the components will be deployed using Kubernetes manifests through Salt
inside a ``metalk8s-logging`` Namespace.

Fluent Bit
""""""""""

Fluent Bit will be deployed as a DaemonSet, because we need one collector on
each node to be able to collect logs from both the Kubernetes platform, the
applications that run on top of it and all the system daemons running alongside

> (e.g. Salt minion).

Reviewer 2: Minor: IMO worth mentioning kubelet also here

Author: Yes, exempli gratia is not intented to be exhaustive, but maybe since Kubelet is a really important component, it could make sens to add it here. I'll do it.

Loki Storage
____________

Since we're using filesystem to store Loki's data, we have basically 2 ways
for having multiple Loki instances running at the same time, with the same
set of data.

Either we do like Prometheus and we store everything on every instance of
Loki, but it means we raise the storage, RAM and CPU needs for each additionnal
instance, either we use a `new experimental feature`_ from Loki 1.5.0 where
Loki ingesters use a hash ring and talk between them to route the queries to
the right one. This approach needs an external KV store to work such as
etcd or Consul.

We choose to use the first approach as the second is not production ready
and since we plan to only do short term retention on Loki, the impact on
storage will not be that much important.
Moreover, it eases the deployment and maintenance since there is not an
extra component.

.. _new experimental feature: https://github.com/grafana/loki/blob/master/docs/operations/storage/filesystem.md#new-and-very-experimental-in-150-horizontal-scaling-of-the-filesystem-store

.. todo::

  define loki storage sizing depending on retention, ingestion rate,

>   average size of logs, number of streams, ...

Reviewer 3: Yes, but make this `todo` item more than just about storage: we need a sizing rule for CPU/RAM as well. This could be stated next to the numbers report you added above.

This configuration will also be customizable by the user to be able to
add new routes (Output) to push the log streams to.

This configuration will be done through various CRs provided by the Fluent Bit
Operator:

 - FluentBit: Defines Fluent Bit instances and its associated config
 - FluentBitConfig: Select input/filter/output plugins and generates the final
   config into a Secret
 - Input: Defines input config sections
 - Filter: Defines filter config sections
 - Output: Defines output config sections

For example, if a user wants to forward all Kubernetes logs to an external log
centralization system (e.g. Elasticisearch), he will need to define an
``Output`` CR as follows:

.. code-block:: yaml

  kind: Output
  metadata:
    name: my-output-to-external-es
    namespace: my-namespace
  spec:
    match: kube.*
    es:
      host: 10.0.0.1
      port: 9200

More details can be found on `Fluent Bit Operator repository`_ and manifest
samples are
`here <https://github.com/kubesphere/fluentbit-operator/tree/master/manifests/logging-stack>`_.

.. _Fluent Bit Operator repository: https://github.com/kubesphere/fluentbit-operator

Loki
""""

> Loki's configuration is stored as a Secret, as for Fluent Bit, we need

Reviewer 2: Minor: Since we use fluent-bit operator it's not "as for Fluent Bit" I think

Author: Yes, good catch, it's a leftover from the previous version of the design.

  define loki storage sizing depending on retention, ingestion rate,
  average size of logs, number of streams, ...

Configuration
~~~~~~~~~~~~~

Fluent Bit
""""""""""

Fluent Bit needs to be configured to scrape and handle properly journal and
containers logs by default.

For containers logs, we want to add the following labels:

 - node: the node it comes from
 - namespace: the namespace the pod is running in
 - instance: the name of the pod
 - container: the name of the container

For journal we want these labels:

 - node: the node it comes from
 - unit: the name of the unit generating these logs

This configuration will also be customizable by the user to be able to
add new routes (Output) to push the log streams to.

This configuration will be done through various CRs provided by the Fluent Bit
Operator:

 - FluentBit: Defines Fluent Bit instances and its associated config
 - FluentBitConfig: Select input/filter/output plugins and generates the final
   config into a Secret
 - Input: Defines input config sections
 - Filter: Defines filter config sections
 - Output: Defines output config sections

For example, if a user wants to forward all Kubernetes logs to an external log

> centralization system (e.g. Elasticisearch), he will need to define an

Reviewer 3: ```suggestion
centralization system (e.g. Elasticsearch), he will need to define an
```

Author: No no, it's Elastickysearch... :D

``Output`` CR as follows:

.. code-block:: yaml

  kind: Output
  metadata:
    name: my-output-to-external-es
    namespace: my-namespace
  spec:
    match: kube.*
    es:
      host: 10.0.0.1
      port: 9200

More details can be found on `Fluent Bit Operator repository`_ and manifest
samples are
`here <https://github.com/kubesphere/fluentbit-operator/tree/master/manifests/logging-stack>`_.

.. _Fluent Bit Operator repository: https://github.com/kubesphere/fluentbit-operator

Loki
""""

Loki's configuration is stored as a Secret.
we need to expose few parameters to the user for customization
(e.g. retention).

> Since we do not have Operator and CRs for Loki, we will use the CSC mechanism

Reviewer 3: But, "we do not have Operator" either for fluent-bit, today... Is there a good reason why:

- we need an Operator and CRs for fluent-bit that CSC cannot solve?
- we can simply rely on CSC for Loki configuration?

As explained offline, IMO the design document is here to explain design choices, not to detail the implementation plan. This plan may seem good right now and obsolete in a few months/years - OTOH, if we say "fluent-bit + loki has been chosen because it has a much lower minimal footprint (fixed cost) and allows better scaling in ingestion rate than other architectures", this has value even if new alternatives more powerful than fluent-bit arrived on the market.

It shows how we want to give priority to certain aspects of the software.

Author: CSC is not as powerful as an Operator (it could if we invest time on it, but from my understanding, we want to have real Operators when it's possible, which allow to watch and reconciliate in real time rather than having to launch salt state after each change).
For Loki, I don't think that's a problem to use CSC, since its configuration should not change that much and should only be edited by the MetalK8s cluster administrator.
But, for Fluent Bit, it's not the same, for example, we may want Solutions to be able to provide CRs to define their own filtering/parsing rules.
I agree, I should probably not talk about the implementation, but from my point of view (which is probably wrong), I think giving some explanations on how we may want to do that allows people to better understand the idea (and we're in implementation details section... what should I write here ?).

Reviewer 3: > CSC is not as powerful as an Operator (it could if we invest time on it, but from my understanding, we want to have real Operators when it's possible, which allow to watch and reconciliate in real time rather than having to launch salt state after each change).

That's granted :)

> For Loki, I don't think that's a problem to use CSC, since its configuration should not change that much and should only be edited by the MetalK8s cluster administrator.
> But, for Fluent Bit, it's not the same, for example, we may want Solutions to be able to provide CRs to define their own filtering/parsing rules.

Exactly. That's what we should see in the design.

> I agree, I should probably not talk about the implementation, but from my point of view (which is probably wrong), I think giving some explanations on how we may want to do that allows people to better understand the idea (and we're in implementation details section... what should I write here ?).

Of course, you're right to discuss implementation. What's most important in this document is not the implementation itself however, but what it implies: how it fits the requirements, why this one is chosen over another one, what are its shortcomings, what are the unknowns.

to provide the interface for customization, with a ConfigMap
``metalk8s-loki-config`` in the ``metalk8s-logging`` Namespace.

A Salt module will be developed to manipulate this object, so the logic
can be kept in only one place.

This module must provide necessary methods to show or edit the configuration
in 2 different ways:

   * simple

>    * advanced

Reviewer 3: The format shown below is `advanced`? If yes, then let's not care about a `simple` mode at all, IMO. As @TeddyAndrieux proposed when we were discussing this on "basic alerting", the simplicity layer can be only implemented in the user interface. Whether the UI is browser-based or terminal-based will change the audience anyway, and as such, what's simple for one isn't for the other.

My suggestion is thus: build the sufficiently advanced configuration layer to get everything working, and see what's needed in the UIs. Of course, using real CRs instead of arbitrary text in ConfigMaps will help to avoid errors when UIs will consume/edit this configuration layer.

    labels:
      grafana_datasource: "1"
  data:
    loki-datasource.yaml: |-
      apiVersion: 1
      datasources:
      - name: Loki
        type: loki
        access: proxy
        url: http://loki.loki.svc.cluster.local:3100
        version: 1

To display the logs we also need a dashboard, adding a ConfigMap
``loki-logs-dashboard`` in Namespace ``metalk8s-monitoring``:

.. code-block:: yaml

  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: loki-logs-dashboard
    namespace: metalk8s-monitoring
    labels:
      grafana_dashboard: "1"
  data:
    loki-logs.json: <DASHBOARD DEFINITION>

For the dashboard we will use a view with a simple log panel and variables
representing labels to filter on.
Since journal and kubernetes logs will not have the same labels, we could
either have 2 distinct dashboards or 2 log panels in the same.
An example of what we want is `Loki dashboard`_.

.. _Loki dashboard: https://github.com/grafana/loki/blob/master/production/loki-mixin/dashboard-loki-logs.json

.. note::

  The ``grafana_datasource: "1"`` and ``grafana_dashboard: "1"`` labels are
  what is used by the Prometheus Operator to retrieve datasource and dashboard

>   for Grafana.

Reviewer 3: I guess it also needs to retrieve them from the same namespace as the Operator runs in (`metalk8s-monitoring`)? In any case, could be worth thinking about those the same way we'll think about Solution dashboards/datasources in the future.

Author: I'm not sure they need to be in `metalk8s-monitoring` namespace, I think the Operator has cluster wide rights allowing it to watch/read ConfigMaps.
But I'm not super familiar with ClusterRole and stuff, so maybe I'm misunderstanding something.

Reviewer 3: In this case, it would make sense to store logging-related dashboards/datasources in `metalk8s-logging`

Author: Tested, it does not seem to work if outside of the `metalk8s-monitoring` namespace, at least for datasource (it may not be the case for dashboards, but I didn't find a clear answer, I should test it).

> Loki Volume Purge

Reviewer 3: One really strong message we received from field experts (cc @xaf-scality @bkettler) is that even though the volume is filling up at a dangerous rate, it should never get to the point where it's full.

Monitoring and alerting are paramount for an admin to make sure the system is stable, but a broken component or a DDoS attack over the week-end could result in ingestion increase, which shouldn't break the log system to a point where complex manual operations are required.

Depending on the cause, the logging pipeline could (should?) first pause automatically, and then:

- storage should be adjusted (as you said, by lowering retention or increasing capacity)
- outliers should be detected and silenced / removed from the pipeline (manually or automatically, depends on what we can do)
- ...other ideas?

Author: I don't think we can go to a point where complex manual operations would be required if the storage is filled up, we just need to manually purge it (and probably restard loki pod).

First, for the pause in the log ingestion, I think it's already made this way in fluent-bit.
If an output plugin is unable to send data to a backend, it'll keep the data in memory and if the buffer is full, then it tells the input plugin to pause, see https://fluentbit.io/documentation/0.13/configuration/backpressure.html for more details.

For the storage resizing, we can't do much except providing some alerting (with prediction and also real time monitoring), documentation (through runbook for example) and good sizing rules.

For the outliers... I'm not sure how to achieve this automatically, it means we need to be able to analyze log flow and detect patterns... I'm not sure that's what we want in our simple logging stack.
To me, the pause + alert + documentation should be enough.
Maybe we can also provide some documentation to explain how to do basic filtering to be sure unwanted logs are dropped at collector level.

Otherwise, we could add an extra option in Loki configuration that would allow us to do purging based on size, which could be done with 2 options (to ensure we will not remove all logs if there is a crazy amount of logs):
- % of space used: we purge oldest logs if space is greater than a certain threshold 
- minimum days of retention: we purge logs older than minimum days of retention
It means, for example, you can have a retention of 2 weeks of logs, with a minimum of 3 days and if it's still not sufficient, fluent-bit will pause the flow.
I just think that if we go with automatic purge based on size, we need to be careful and trigger alerts each time we purge, otherwise operators will think they have 2 weeks of logs and will never be aware that there is a sizing issue on their platform.

~~~~~~~~~~~~~~~~~

Even with a max retention period, the logs could grow faster than what was
expected and fill up the volume.
Since there is no retention based on size in Loki yet, we need to add some
specific monitoring (with prediction on volume usage) and alerting to ensure
that an administrator will be warned if such a case would happen.
The alert message should be clear and provide an URL to a run book to help
the administrator resolving the issue.

To fix this issue, the administrator should purge oldest log chunk files
from Loki volume, which can be achieved by connecting to the pod and manually
removing them.
If the growth of logs is not something transient, the administrator should
also be advised to lower the retention period or replace the Loki volume
by a bigger one.

Iterations
~~~~~~~~~~

Iteration 1
"""""""""""

The goal is to have a working log centralization system, with logs accessible
from Grafana:

* Deploy Fluent Bit and Loki
* Deploy Grafana datasource & dashboard

> * Deploy Loki volumes purge mechanism

Reviewer 3: ? Which mechanism?

Author: No mechanism now, it's a leftover.
It's more about having dedicated alert rules for Loki volume usage, I'll update this.
But maybe we still want to have something, depending on what we decide regarding your comment above (https://github.com/scality/metalk8s/pull/2668#discussion_r464871732)

> * Customization of Loki with CSC mechanisms

Reviewer 3: IMO, first iteration should be minimal (just as Dex currently is). We gather the required features from other squads and focus on this. Future iterations will add fields to the available CSC(s).

Author: Finally added it back as it does not cost much to do (and it's already done :p).

The Post Installation page is updated to indicate that persistent storage is
needed for log centralization service.

> A new page should be added to explains how to operate the service and how to

Reviewer 3: ```suggestion
A new page should be added to explain how to operate the service and how to
```
